cid-fmt/main.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+
+	c "github.com/ipfs/go-cid"
+
+	mb "github.com/multiformats/go-multibase"
+	mh "github.com/multiformats/go-multihash"
+)
+
+func usage() {
+	fmt.Fprintf(os.Stderr, "usage: %s [-b multibase-code] <fmt-str> <cid> ...\n\n", os.Args[0])
+	fmt.Fprintf(os.Stderr, "<fmt-str> is either 'prefix' or a printf style format string:\n%s", fmtRef)
+	os.Exit(1)
+}
+
+const fmtRef = `
+   %% literal %
+   %b multibase name 
+   %B multibase code
+   %v version string
+   %V version number
+   %c codec name
+   %C codec code
+   %h multihash name
+   %H multihash code
+   %L hash digest length
+   %m multihash encoded in base %b (with multibase prefix)
+   %M multihash encoded in base %b without multibase prefix
+   %d hash digest encoded on base %b (with multibase prefix)
+   %D hash digest encoded on base %b without multibase prefix
+   %s cid string encoded in base %b (1)
+   %s cid string encoded in base %b without multibase prefix
+   %P cid prefix: %v-%c-%h-%L
+
+(1) For CID version 0 the multibase must be base58btc and no prefix is
+used.  For Cid version 1 the multibase prefix is included.
+`
+
+func main() {
+	if len(os.Args) < 2 {
+		usage()
+	}
+	newBase := mb.Encoding(-1)
+	args := os.Args[1:]
+	if args[0] == "-b" {
+		if len(args) < 2 {
+			usage()
+		}
+		if len(args[1]) != 1 {
+			fmt.Fprintf(os.Stderr, "Error: Invalid multibase code: %s\n", args[1])
+		}
+		newBase = mb.Encoding(args[1][0])
+		args = args[2:]
+	}
+	if len(args) < 2 {
+		usage()
+	}
+	fmtStr := args[0]
+	switch fmtStr {
+	case "prefix":
+		fmtStr = "%P"
+	default:
+		if strings.IndexByte(fmtStr, '%') == -1 {
+			fmt.Fprintf(os.Stderr, "Error: Invalid format string: %s\n", fmtStr)
+		}
+	}
+	for _, cidStr := range args[1:] {
+		base, cid, err := decode(cidStr)
+		if newBase != -1 {
+			base = newBase
+		}
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s: %v\n", cidStr, err)
+			fmt.Fprintf(os.Stdout, "!INVALID_CID!\n")
+			// Don't abort on a bad cid
```

Comment:
I can return non-zero exit status.  But if multiple CID are given it may be useful to try and convert the other arguments.